Doc/reference/datamodel.rst

```diff
@@ -2037,12 +2037,15 @@ Basic customization
    .. index:: single: __len__() (mapping object method)
 
    Called to implement truth value testing and the built-in operation
-   ``bool()``; should return ``False`` or ``True``.  When this method is not
+   :func:`bool`; should return ``False`` or ``True``.  When this method is not
    defined, :meth:`~object.__len__` is called, if it is defined, and the object is
    considered true if its result is nonzero.  If a class defines neither
    :meth:`!__len__` nor :meth:`!__bool__`, all its instances are considered
    true.
 
+   Two successive calls to :meth:`!__bool__` on the same object must
+   return same value.
```

Comment:
Not if the object is mutated.

Reply:
But that means - some other method was called on the object in between.  Thus, calls to ``__bool__`` weren't actually successive.

Reply:
The object could have been mutated by another thread between the two calls...

Reply:
> by another thread

... that calls some other object's method

> between the two calls...

:)

Reply:
Hmm, how about this: "The ``__bool__`` method can't mutate the object."?  (Ditto for ``__len__``.)

This probably is a more strong requirement than actually need in current optimizations, but I doubt it blocks something useful.